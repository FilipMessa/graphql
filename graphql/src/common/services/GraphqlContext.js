// @flow

import DataLoader from 'dataloader';

import IdentityDataloader from '../../identity/dataloaders/Identity';
import createBookingLoader from '../../booking/dataloaders/Booking';
import createAirlineLoader from '../../flight/dataloaders/Airline';
import createRatesLoader from '../dataloaders/Rates';
import BookingsLoader from '../../booking/dataloaders/Bookings';
import LocationSuggestionsLoader from '../../location/dataloaders/LocationSuggestions';
import LocationLoader from '../../location/dataloaders/Location';
import FlightLoader from '../../flight/dataloaders/Flight';
import OptionsStorage from './context/OptionsStorage';
import HotelsAvailability from '../../hotel/dataloaders/HotelsAvailability';
import { type SearchParameters as HotelKey } from '../../hotel/dataloaders/flow/SearchParameters';
import HotelByID from '../../hotel/dataloaders/HotelByID';
import HotelCities from '../../hotel/dataloaders/HotelCities';
import HotelRoomsLoader from '../../hotel/dataloaders/HotelRooms';
import HotelRoomAvailabilityLoader from '../../hotel/dataloaders/HotelRoomAvailability';
import HotelRoomBeddingLoader from '../../hotel/dataloaders/RoomBedding';
import PriceStatsLoader from '../../hotel/dataloaders/PriceStats';

import type { Booking } from '../../booking/Booking';
import type { Airline } from '../../flight/Flight';
import type { HotelType } from '../../hotel/dataloaders/flow/HotelType';
import type { HotelCity } from '../../hotel/types/outputs/HotelCity';

/**
 * FIXME: this is stupid - it's already defined by data-loader itself
 */
export type GraphqlContextType = {|
  // DataLoader<K, V>
  apiToken: ?string,
  dataLoader: {|
    airline: DataLoader<string, ?Airline>,
    booking: DataLoader<number | string, Booking>,
    bookings: BookingsLoader,
    flight: FlightLoader,
    identity: IdentityDataloader,
    location: LocationLoader,
    locationSuggestions: LocationSuggestionsLoader,
    rates: DataLoader<string, ?number>,
    hotel: {
      availabilityByLocation: DataLoader<HotelKey, HotelType[]>,
      availabilityByID: DataLoader<HotelKey, HotelType[]>,
      byID: typeof HotelByID,
      cities: DataLoader<string, HotelCity[]>,
      room: HotelRoomsLoader,
      roomAvailability: HotelRoomAvailabilityLoader,
      roomBedding: typeof HotelRoomBeddingLoader,
      priceStats: DataLoader<
        { searchParams: HotelKey, boundary: 'MAX' | 'MIN' },
        number,
      >,
    },
  |},
  options: OptionsStorage,
  _traceCollector?: Object,
|};

export function createContext(token: ?string): GraphqlContextType {
  const bookings = new BookingsLoader(token);
  const locationSuggestions = new LocationSuggestionsLoader();
  const location = new LocationLoader(locationSuggestions);

  return {
    apiToken: token,
    dataLoader: {
      airline: createAirlineLoader(),
      booking: createBookingLoader(token, bookings),
      bookings: bookings,
      flight: new FlightLoader(location),
      identity: new IdentityDataloader(token),
      location: location,
      locationSuggestions: locationSuggestions,
      rates: createRatesLoader(),
      hotel: {
        availabilityByLocation: HotelsAvailability,
        availabilityByID: HotelsAvailability,
        byID: HotelByID,
        cities: HotelCities,
        room: new HotelRoomsLoader(),
        roomAvailability: new HotelRoomAvailabilityLoader(),
        roomBedding: HotelRoomBeddingLoader,
        priceStats: PriceStatsLoader,
      },
    },
    options: new OptionsStorage(),
  };
}
