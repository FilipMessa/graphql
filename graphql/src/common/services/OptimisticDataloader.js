// @flow

import DataLoader from 'dataloader';

// any because it must be very generic (accepts everything)
type K = any;
type V = any;

/**
 * Original data-loader kills load batch if one of this response values is
 * Error instance. This optimistic data-loader keeps Error value as a valid
 * response. This means that it's possible to return as many responses
 * as possible and convert Errors to the error messages in the response
 * if necessary.
 *
 * @see https://github.com/facebook/dataloader/issues/41
 */
export default class OptimisticDataloader extends DataLoader<K, V> {
  constructor(batchLoadFn: Function) {
    super(batchLoadFn);
  }

  loadMany = (keys: K[]): Promise<V[]> => {
    if (!Array.isArray(keys)) {
      throw new TypeError(
        'The loader.loadMany() function must be called with Array<key> ' +
          `but got: ${keys}.`,
      );
    }

    return Promise.all(
      keys.map(key =>
        this.load(key).catch(error => {
          if (error instanceof TypeError) {
            // original data-loader throws TypeErrors
            throw error;
          }
          return error;
        }),
      ),
    );
  };
}
