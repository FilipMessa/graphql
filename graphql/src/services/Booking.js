import jsonMock from '../../tests/bookings.json';

// const url = `https://booking-api.skypicker.com/api/v0.1/users/self/bookings?token=${token}`;

export default () => jsonMock; // FIXME: replace with API call
