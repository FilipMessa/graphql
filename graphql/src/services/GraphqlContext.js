// @flow

import DataLoader from 'dataloader';
import createIdentityLoader from '../dataLoaders/Identity';
import createBookingLoader from '../dataLoaders/Booking';
import createAirlineLoader from '../dataLoaders/Airline';
import BookingsLoader from '../dataLoaders/Bookings';
import type { IdentityType, BookingType, AirlineType } from '../Entities';

export type GraphqlContextType = {
  apiToken: ?string,
  dataLoader: {
    booking: DataLoader<number | string, BookingType>,
    bookings: BookingsLoader,
    identity: DataLoader<string, IdentityType>,
    airline: DataLoader<string, ?AirlineType>,
  },
};

export function createContext(token: ?string): GraphqlContextType {
  const bookings = new BookingsLoader(token);
  return {
    apiToken: token,
    dataLoader: {
      booking: createBookingLoader(token, bookings),
      bookings: bookings,
      identity: createIdentityLoader(token),
      airline: createAirlineLoader(),
    },
  };
}
