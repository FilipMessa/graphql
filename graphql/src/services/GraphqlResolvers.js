// @flow

import _ from 'lodash';
import differenceInMinutes from 'date-fns/difference_in_minutes';
import type { ArrivalType, DepartureType } from '../Entities';

export function flightDurationInMinutes(
  departure: DepartureType,
  arrival: ArrivalType,
): ?number {
  const departureTime = _.get(departure, 'when.utc');
  const arrivalTime = _.get(arrival, 'when.utc');
  return dateDurationInMinutes(departureTime, arrivalTime);
}

export function dateDurationInMinutes(departure: ?Date, arrival: ?Date) {
  if (!departure || !arrival || departure > arrival) {
    return null;
  }
  return differenceInMinutes(arrival, departure);
}
