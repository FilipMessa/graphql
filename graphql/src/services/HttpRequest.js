// @flow

import url from 'url';
import fetch from 'node-fetch';
import Logger from '../services/Logger';
import { ProxiedError } from './errors/ProxiedError';

export default function request(
  absoluteApiUrl: string,
  token: ?string,
): Promise<Object> {
  const urlObject = url.parse(absoluteApiUrl, true);
  if (token !== null && token !== undefined) {
    if (urlObject.query === undefined) {
      urlObject.query = {};
    }
    urlObject.query.token = token;
  }

  return fetchJson(url.format(urlObject));
}

export async function post(
  absoluteApiUrl: string,
  payload: Object,
  requestHeaders?: Object = {},
): Promise<Object> {
  const body = JSON.stringify(payload);
  const headers = {
    'Content-Type': 'application/json',
    ...requestHeaders,
  };

  return fetchJson(absoluteApiUrl, 'POST', { body, headers });
}

async function fetchJson(
  url: string,
  method: string = 'GET',
  options?: Object = {},
): Promise<Object> {
  if (process.env.NODE_ENV === 'test') {
    throw new Error(
      `HttpRequest should never be called in test environment. Have you forgotten to mock "${url}" with fake data response?`,
    );
  }
  if (typeof url !== 'string') {
    throw new Error(
      `Relative API URL should be typeof 'string', '${typeof url}' given.`,
    );
  }

  const reqOptions = {
    ...options,
    method,
  };

  Logger.info(`${method} ${url}`);
  const response = await fetch(url, reqOptions);

  if (response.status !== 200) {
    throw new ProxiedError(response.statusText, response.status, url);
  }
  return response.json();
}
