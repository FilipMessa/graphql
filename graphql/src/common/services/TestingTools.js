// @flow

import { graphql as graphqlNodeLibrary } from 'graphql';
import schema from '../../Schema';
import { createContext } from './GraphqlContext';

/**
 * TestingTools must be imported FIRST otherwise mock of HttpRequest won't work.
 */
jest.mock('../services/HttpRequest');
const mockedHttpRequest = require('./HttpRequest');

class MockResponse {
  absoluteUrl: string;
  httpMethod: string;

  constructor(httpMethod: 'GET' | 'POST', absoluteUrl: string) {
    this.httpMethod = httpMethod;
    this.absoluteUrl = absoluteUrl;
  }

  replyWithData(response: Array<Object> | Object) {
    // $FlowAllowMockMagic
    mockedHttpRequest.__setMockData(
      this.httpMethod,
      this.absoluteUrl,
      response,
    );
  }
}

export class RestApiMock {
  static onGet(absoluteUrl: string) {
    return new MockResponse('GET', absoluteUrl);
  }

  static onPost(absoluteUrl: string) {
    return new MockResponse('POST', absoluteUrl);
  }
}

export const graphql = async (
  query: string,
  variables: ?Object,
): Promise<Object> =>
  graphqlNodeLibrary(
    schema,
    query,
    null,
    createContext('test_token'),
    variables,
  );

export const evaluateResolver = (field: mixed, testValue: mixed) => {
  // $FlowAllowNextLineInThisTest (fields are possibly undefined but we assume it's fine)
  return field.resolve(testValue);
};
