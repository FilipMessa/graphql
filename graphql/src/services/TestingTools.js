import { graphql } from 'graphql';
import schema from '../Schema';
import { createContext } from '../services/GraphqlContext';

jest.mock('../services/HttpRequest');

export const executeQuery = async (query: string): Promise<Object> =>
  graphql(schema, query, null, createContext('test_token'));
