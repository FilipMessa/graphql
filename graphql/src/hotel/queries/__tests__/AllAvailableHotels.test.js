// @flow

import { graphql, RestApiMock } from '../../../common/services/TestingTools';
import AllHotelsDataset from '../../datasets/all.json';
import FirstHotelDataset from '../../datasets/25215.json';
import SecondHotelDataset from '../../datasets/248539.json';
import RoomBlocksDataset from '../../datasets/roomBlocks.json';
import RoomDetailsDataset from '../../datasets/roomDetails.json';

beforeEach(() => {
  // minimal query
  RestApiMock.onGet(
    'https://hotels-api.skypicker.com/api/hotels?output=hotel_details&radius=50&latitude=45.4654&longitude=9.1859&checkin=2017-11-16&checkout=2017-11-23&room1=A',
  ).replyWithData(AllHotelsDataset);

  // full query
  RestApiMock.onGet(
    'https://hotels-api.skypicker.com/api/hotels?output=hotel_details&radius=50&latitude=45.4654&longitude=9.1859&checkin=2017-11-16&checkout=2017-11-23&room1=A%2CA%2C4%2C6&room2=A%2C2',
  ).replyWithData(AllHotelsDataset);
  RestApiMock.onGet(
    'https://hotels-api.skypicker.com/api/hotelDetails?hotel_ids=25215%2C248539',
  ).replyWithData(FirstHotelDataset.concat(SecondHotelDataset));
  RestApiMock.onGet(
    'https://hotels-api.skypicker.com/api/blocks?arrival_date=2017-11-16&departure_date=2017-11-23&hotel_ids=25215',
  ).replyWithData(RoomBlocksDataset);
  RestApiMock.onGet(
    // this can be merged into previous request (hotel_ids=25215,248539) but we are not doing this optimization yet
    'https://hotels-api.skypicker.com/api/blocks?arrival_date=2017-11-16&departure_date=2017-11-23&hotel_ids=248539',
  ).replyWithData(RoomBlocksDataset);
  RestApiMock.onGet(
    'https://hotels-api.skypicker.com/api/roomDetails?hotel_ids=25215',
  ).replyWithData(RoomDetailsDataset);
  RestApiMock.onGet(
    'https://hotels-api.skypicker.com/api/roomDetails?hotel_ids=248539',
  ).replyWithData([]); // Booking.com failure
});

describe('all hotels query', () => {
  it('should work for minimal example', async () => {
    expect(
      await graphql(`
        {
          allAvailableHotels(
            search: {
              latitude: 45.4654
              longitude: 9.1859
              checkin: "2017-11-16"
              checkout: "2017-11-23"
              roomsConfiguration: { adultsCount: 1 }
            }
          ) {
            edges {
              cursor
              node {
                id
              }
            }
          }
        }
      `),
    ).toMatchSnapshot();
  });

  it('should work for full query', async () => {
    expect(
      await graphql(`
        {
          allAvailableHotels(
            search: {
              latitude: 45.4654
              longitude: 9.1859
              checkin: "2017-11-16"
              checkout: "2017-11-23"
              roomsConfiguration: [
                { adultsCount: 2, children: [{ age: 4 }, { age: 6 }] }
                { adultsCount: 1, children: [{ age: 2 }] }
              ]
            }
          ) {
            edges {
              node {
                id
                price {
                  amount
                  currency
                }
                availableRooms {
                  id
                  originalId
                  minimalPrice {
                    amount
                    currency
                  }
                  incrementalPrice {
                    amount
                    currency
                  }
                  room {
                    id
                    type
                  }
                }
                hotel {
                  id
                  originalId
                  summary
                }
              }
            }
          }
        }
      `),
    ).toMatchSnapshot();
  });
});
