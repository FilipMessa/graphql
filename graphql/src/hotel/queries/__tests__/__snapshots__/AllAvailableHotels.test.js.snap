// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all hotels query should work for full query 1`] = `
Object {
  "data": Object {
    "allAvailableHotels": Object {
      "edges": Array [
        Object {
          "node": Object {
            "availableRooms": Array [
              Object {
                "id": "SG90ZWxSb29tQXZhaWxhYmlsaXR5OjI0ODUzOTAxXzk0MzY0OTI4XzFfMl8w",
                "incrementalPrice": Array [
                  Object {
                    "amount": 120,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 240,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 360,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 480,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 600,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 720,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 840,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 960,
                    "currency": "EUR",
                  },
                ],
                "minimalPrice": Object {
                  "amount": 120,
                  "currency": "EUR",
                },
                "room": null,
              },
              Object {
                "id": "SG90ZWxSb29tQXZhaWxhYmlsaXR5OjI0ODUzOTAxXzEwNjY2NjM4N18xXzJfMA==",
                "incrementalPrice": Array [
                  Object {
                    "amount": 127,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 254,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 381,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 508,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 635,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 762,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 889,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 1016,
                    "currency": "EUR",
                  },
                ],
                "minimalPrice": Object {
                  "amount": 127,
                  "currency": "EUR",
                },
                "room": null,
              },
              Object {
                "id": "SG90ZWxSb29tQXZhaWxhYmlsaXR5OjI1MjE1MTVfOTUxNDkwMzZfMF8xXzA=",
                "incrementalPrice": Array [
                  Object {
                    "amount": 62.1,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 124.2,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 186.3,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 248.4,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 310.5,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 372.6,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 434.7,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 496.8,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 558.9,
                    "currency": "EUR",
                  },
                ],
                "minimalPrice": Object {
                  "amount": 62.1,
                  "currency": "EUR",
                },
                "room": Object {
                  "id": "aG90ZWxSb29tOjI1MjE1MTU=",
                  "type": "Twin/Double",
                },
              },
              Object {
                "id": "SG90ZWxSb29tQXZhaWxhYmlsaXR5OjI1MjE1MTVfMTAzMDE5NjI5XzBfMV8w",
                "incrementalPrice": Array [
                  Object {
                    "amount": 69,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 138,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 207,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 276,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 345,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 414,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 483,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 552,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 621,
                    "currency": "EUR",
                  },
                ],
                "minimalPrice": Object {
                  "amount": 69,
                  "currency": "EUR",
                },
                "room": Object {
                  "id": "aG90ZWxSb29tOjI1MjE1MTU=",
                  "type": "Twin/Double",
                },
              },
            ],
            "hotel": Object {
              "id": "aG90ZWw6MjUyMTU=",
            },
            "id": "SG90ZWxBdmFpbGFiaWxpdHk6MjUyMTU=",
            "price": Object {
              "amount": 1134.9,
              "currency": "EUR",
            },
          },
        },
        Object {
          "node": Object {
            "availableRooms": Array [
              Object {
                "id": "SG90ZWxSb29tQXZhaWxhYmlsaXR5OjI0ODUzOTAxXzk0MzY0OTI4XzFfMl8w",
                "incrementalPrice": Array [
                  Object {
                    "amount": 120,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 240,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 360,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 480,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 600,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 720,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 840,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 960,
                    "currency": "EUR",
                  },
                ],
                "minimalPrice": Object {
                  "amount": 120,
                  "currency": "EUR",
                },
                "room": null,
              },
              Object {
                "id": "SG90ZWxSb29tQXZhaWxhYmlsaXR5OjI0ODUzOTAxXzEwNjY2NjM4N18xXzJfMA==",
                "incrementalPrice": Array [
                  Object {
                    "amount": 127,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 254,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 381,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 508,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 635,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 762,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 889,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 1016,
                    "currency": "EUR",
                  },
                ],
                "minimalPrice": Object {
                  "amount": 127,
                  "currency": "EUR",
                },
                "room": null,
              },
              Object {
                "id": "SG90ZWxSb29tQXZhaWxhYmlsaXR5OjI1MjE1MTVfOTUxNDkwMzZfMF8xXzA=",
                "incrementalPrice": Array [
                  Object {
                    "amount": 62.1,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 124.2,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 186.3,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 248.4,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 310.5,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 372.6,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 434.7,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 496.8,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 558.9,
                    "currency": "EUR",
                  },
                ],
                "minimalPrice": Object {
                  "amount": 62.1,
                  "currency": "EUR",
                },
                "room": Object {
                  "id": "aG90ZWxSb29tOjI1MjE1MTU=",
                  "type": "Twin/Double",
                },
              },
              Object {
                "id": "SG90ZWxSb29tQXZhaWxhYmlsaXR5OjI1MjE1MTVfMTAzMDE5NjI5XzBfMV8w",
                "incrementalPrice": Array [
                  Object {
                    "amount": 69,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 138,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 207,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 276,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 345,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 414,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 483,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 552,
                    "currency": "EUR",
                  },
                  Object {
                    "amount": 621,
                    "currency": "EUR",
                  },
                ],
                "minimalPrice": Object {
                  "amount": 69,
                  "currency": "EUR",
                },
                "room": Object {
                  "id": "aG90ZWxSb29tOjI1MjE1MTU=",
                  "type": "Twin/Double",
                },
              },
            ],
            "hotel": Object {
              "id": "aG90ZWw6MjQ4NTM5",
            },
            "id": "SG90ZWxBdmFpbGFiaWxpdHk6MjQ4NTM5",
            "price": Object {
              "amount": 2471.98,
              "currency": "EUR",
            },
          },
        },
      ],
    },
  },
}
`;

exports[`all hotels query should work for minimal example 1`] = `
Object {
  "data": Object {
    "allAvailableHotels": Object {
      "edges": Array [
        Object {
          "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
          "node": Object {
            "id": "SG90ZWxBdmFpbGFiaWxpdHk6MjUyMTU=",
          },
        },
        Object {
          "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
          "node": Object {
            "id": "SG90ZWxBdmFpbGFiaWxpdHk6MjQ4NTM5",
          },
        },
      ],
    },
  },
}
`;
